### ISSUE TITLE
support comparison checks using Comparator instead of Comparable

### ISSUE THREAD

**Issue by HermanBovens** [assertj/assertj-core#2041]
#### Summary
Currently we have a fluent API for asserting comparison results: `Assertions.assertThat(java.lang.Comparable)`, where the ordering can also be replaced by a custom comparator by the use of `usingComparator(java.lang.Comparator)`.
I'd like to have the same API available through the use of `Assertions.assertThat(Object).usingComparator(java.lang.Comparator)` for classes that don't implement `Comparable`.

#### Example

```java

assertThat(object1).usingComparator(mySpecialOrdering).isLessThan(object2);

```

**Comment by joel-costigliola** [comment]
I'm sympathetic to the idea but with the current design this is not really possible, `usingComparator` returns the current assertions object which allows specific assertions on the the object under test actual type, to provide comparable assertions we would have to return a comparable assertion object but then you loose the specific assertions type of the object under test.
Quite a big breaking change.

**Comment by scordio** [comment]
I also don't see an easy way to have both type-specific assertions and comparable assertions for an object which doesn't implement `Comparable`. If we are willing to sacrifice the type-specific assertions, we could offer another entry point in `Assertions` like:

```java
public static <T> AbstractComparableAssert<?, ?> assertThat(T actual, Comparator<T> comparator) {
  ...
}
```

The challenge might be how to construct the comparable assert instance, we probably need an adapter to use an object + comparator as a comparable instance.

@HermanBovens would this approach help? If you like the idea, could you share a more detailed example?

**Comment by HermanBovens** [comment]
I guess something like that would help.  Or alternatively, a different method than `usingComparator`, which already exists.  For example:

```
assertThat(o1).comparableUsing(Comparator<T> comparator).isGreaterThan(o2)
```

The need to construct the comparable assert instance stays the same.  Maybe an adapter would work, but that doesn't sound natural.  For me it would be far more logical to implement the general case using a comparator, and for comparable assertions simply use that implementation but with `Comparators.naturalOrder()` by default.  That will probably need some refactoring, I don't know the AssertJ code enough to decide if that's a lot of work and if it would be necessarily a breaking change.  I'd think the existing `AbstractComparableAssert` could just delegate to some new `AbstractComparatorAssert`.

The other thing is, when you need to assert a bunch of things all with the same comparator, it's not so great that you have to specify the comparator n times.  Maybe that's a separate feature request, but it would be nice if there was a way to specify it only once and do a bunch of assertions using it.

**Comment by scordio** [comment]
> Or alternatively, a different method than `usingComparator`, which already exists. For example:
> 
> ```
> assertThat(o1).comparableUsing(Comparator<T> comparator).isGreaterThan(o2)
> ```

If I understand it correctly this would be exposed for all the objects, my feeling is it might create confusion being together with `usingComparator(Comparator<? super ACTUAL> customComparator)`.

> Maybe an adapter would work, but that doesn't sound natural. For me it would be far more logical to implement the general case using a comparator, and for comparable assertions simply use that implementation but with `Comparators.naturalOrder()` by default.

I wouldn't touch `AbstractComparableAssert` which does a decent job for `Comparable` objects.

What we are trying to design is beyond the normal behavior as we want to have comparable assertions for non `Comparable` objects, therefore I favor the encapsulation of any needed "ugliness" in something special (i.e., `Assertions.assertThat(T, Comparator<T>)`) but still keeping an understandable and easy to use API.

> The other thing is, when you need to assert a bunch of things all with the same comparator, it's not so great that you have to specify the comparator n times. Maybe that's a separate feature request, but it would be nice if there was a way to specify it only once and do a bunch of assertions using it.

You probably mean asserting different objects without the need of setting the comparator after each `assertThat()`.

I don't see a way to do it at the moment, maybe we could introduce something similar to `Assertions.useRepresentation()`. @joel-costigliola thoughts?

**Comment by joel-costigliola** [comment]
I agree with @scordio on `comparableUsing` being too confusing, no clear indication on the name itself what would be the difference with `usingComparator` and on not changing `AbstractComparableAssert`.

That leaves us with the `assertThat(T, Comparator<T>)` option or thinking out loud `assertThatWithComparator(T)` that would return a specific class similar to `ComparableAssert` but without requiring actual to be comparable.

```java
assertThatWithComparator(person).usingComparator((p1, p2) -> ...)
                                .isLessThan(otherPerson);
```

We could add a `Assertions.registerComparable(Comparator<?>, Class<?> type)` that would be used by `assertThatWithComparator`, ex:

```
Assertions.registerComparable((p1, p2) -> ..., Person.class);
assertThatWithComparator(person).isLessThan(otherPerson);
```

That's some early thoughts, we need to spike the different options before reaching to a decision.

**Comment by HermanBovens** [comment]
>>Maybe an adapter would work, but that doesn't sound natural. For me it would be far more logical to implement the general case using a comparator, and for comparable assertions simply use that implementation but with Comparators.naturalOrder() by default.
>
>I wouldn't touch AbstractComparableAssert which does a decent job for Comparable objects.
>
>What we are trying to design is beyond the normal behavior as we want to have comparable assertions for non Comparable objects, therefore I favor the encapsulation of any needed "ugliness" in something special (i.e., Assertions.assertThat(T, Comparator<T>)) but still keeping an understandable and easy to use API.

I was just thinking about maintenance here.  All the `AbstractComparableAssert` methods do, is verify an ordering.  An ordering can be the natural ordering, defined by the implementation of `Comparable<T>`, or any other ordering, as defined by the implementation of `Comparator<T>`.   There is a static method [naturalOrder()](https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/util/Comparator.html#naturalOrder()) that returns a comparator for `Comparable`s to sort according to their natural ordering.  So you can have all assertion code in one place e.g. in a super class of `AbstractComparableAssert` which would then just call the super constructor using `naturalOrder()`.  So no need to maintain two different but very similar functionalities.

I quickly looked in the codebase and I see you already have this for a large part: there is interface `ComparisonStrategy` that defines the checks and does not depend on `Comparable` in any way.  There is an implementing class `ComparatorBasedComparisonStrategy` which also does not depend on `Comparable`.  You have a `Comparables` object with reusable assertions and which takes a `ComparisonStrategy` and delegates to this.  The only case where the `Comparable` interface is actually used, is in case of `assertEqualByComparison` and `assertNotEqualByComparison`.  If you really want to keep this you can at least change the generic type parameter from `<T extends Comparable<? super T>>` to just `<T>` for the other assertions.  I'd personally change it to always use the comparison strategy (and update the generic type parameter also for `assertEqualByComparison` and `assertNotEqualByComparison`).  That's a breaking change only if these two methods are used after `usingComparator`, I'd be surprised if that is often the case.  You'd of course need to update the javadoc - btw the javadoc of `usingComparator()` is currently incorrect: it only says the `Comparator<T>` will be used for equals comparison, but other assertions such as `isLessThan` also use it.  Currently it should say it is used for all comparable assertions except `assertEqualByComparison` and `assertNotEqualByComparison`.

I agree about the part of `comparableUsing` to be too confusing.  Thinking out loud again: we already have `usingComparator` which is currently only used to override equality checks (so only using the case where `compareTo()` returns 0).  I wouldn't mind if all order-verifying methods from `AbstractComparableAssert` would be available on `AbstractObjectAssert`, and the test would fail for non-`Comparable`s if no `Comparator` was set.  Yes, informing the test developer of her mistake is deferred until runtime, but it may be the simplest in terms of implementation and usage, and building on what already exists.

Regarding the suggestion of `registerComparable`: as it is a static method, wouldn't we have the problem that it affects tests that come after the current test method?  I think you'd need to wrap the actual assertions in a lambda, register the comparator before executing the lambda and unregister it afterwards.  But even then, tests can not be run in parallel (but I'm not sure if AssertJ supports this currently anyway) unless if the comparator would be registered in a `ThreadLocal`.

**Comment by scordio** [comment]
Hi @HermanBovens, I plan to start working on this, so at first I wanted to catch up with your last message.

> the javadoc of `usingComparator()` is currently incorrect: it only says the `Comparator<T>` will be used for equals comparison, but other assertions such as `isLessThan` also use it. Currently it should say it is used for all comparable assertions except `assertEqualByComparison` and `assertNotEqualByComparison`.

Which javadoc are you referring to? The one in `Assert` has just an example showing the usage with `isEqualTo`, but it is not mentioning that `isEqualTo` is the only place where the comparator is used.

>  I wouldn't mind if all order-verifying methods from `AbstractComparableAssert` would be available on `AbstractObjectAssert`, and the test would fail for non-`Comparable`s if no `Comparator` was set. Yes, informing the test developer of her mistake is deferred until runtime, but it may be the simplest in terms of implementation and usage, and building on what already exists.

Having an API that fails under special conditions could confuse the end user. I don't mind paying the price of internal complexity as long as we provide an easy-to-understand user experience and the runtime error introduces additional complexity that maybe we can avoid for the time being.

> Regarding the suggestion of `registerComparable`: as it is a static method, wouldn't we have the problem that it affects tests that come after the current test method?

I understood this is what you wanted to have, but maybe I was wrong. You wrote:

> The other thing is, when you need to assert a bunch of things all with the same comparator, it's not so great that you have to specify the comparator n times. Maybe that's a separate feature request, but it would be nice if there was a way to specify it only once and do a bunch of assertions using it.

Would you mind providing a small example of the APIs or steps you would expect?

**Comment by HermanBovens** [comment]
Hi @scordio,

>> the javadoc of usingComparator() is currently incorrect: it only says the Comparator<T> will be used for equals comparison, but other assertions such as isLessThan also use it. Currently it should say it is used for all comparable assertions except assertEqualByComparison and assertNotEqualByComparison.
>
> Which javadoc are you referring to? The one in Assert has just an example showing the usage with isEqualTo, but it is not mentioning that isEqualTo is the only place where the comparator is used.

The full javadoc in Assert, excluding the code example, reads:

```
Use the given custom comparator instead of relying on actual type A equals method for incoming assertion checks.

The custom comparator is bound to assertion instance, meaning that if a new assertion instance is created, the default
comparison strategy will be used.
```

To me this suggests it only impacts equality checks.  It is incomplete at least.  I had to dig into the code to find out it's also about comparison.

> Having an API that fails under special conditions could confuse the end user. I don't mind paying the price of internal complexity as long as we provide an easy-to-understand user experience and the runtime error introduces additional complexity that maybe we can avoid for the time being.

The "special  condition" would mean that you're trying to trying to assert comparison logic without specifying a `Comparator` or using a `Comparable`.  I think a decent failure message should clear up any confusion easily.

What I find more confusing, is the unnecessary coupling of `Comparable` and `Comparator` conceprts.  In the Java Collections API, once you decide to use a `Comparator` for comparison, it's completely irrelevant whether or not the type you're comparing implements `Comparable`.  The API is always the same.  In other words, `Comparable` and `Comparator` are orthogonal concepts.  So I'd expect this to be the case in other libraries such as AssertJ.  Right now with `usingComparator` that's not the case as the `Comparator` impacts only equality checks for non-`Comparable` assertions but also comparison checks for `Comparable` assertions.  I believe this inconsistency is also what lead to the javadoc being incomplete.

Now, I have been thinking more about this: essentially, when using a `Comparator` to assert comparison logic, the object under test is actually the comparator itself, not the things you're comparing.  So it may make more sense to have the API like this:

```
assertThat(comparator).comparing(a, b).isLessThan();
assertThat(comparator).comparing(a, a).isEqual();
assertThat(comparator).comparing(b, a).isGreaterThan();
```
or even
```
assertThat(comparator).comparing(a, b, c, d, e).isIncreasingOrder();
assertThat(comparator).comparing(e, d, c, b, a).isDecreasingOrder();
```
Note that in this example the order would have to be checked by comparing each element with the next ones, not by checking the result of a "sort" operation: sorting algorithms rely on order being _transitive_ (a < b and b < c means a < c) while a buggy comparator may not have this property, and as a result an incorrectly implemented comparator could yield the correct sort order.

With such an API it wouldn't matter if the type of `a` and `b` implements `Comparable` or not, that's irrelevant as expected.

You could argue that the following is currently already possible:
```
assertThat(comparator.compare(a, b)).isNegative();
assertThat(comparator.compare(a, a)).isZero();
assertThat(comparator.compare(a, b)).isPositive();
```
but the first examples above are more "fluent" (which I think is what AssertJ is all about), and there is no alternative for the second examples other than asserting the comparison result for all successive elements separately.

I'd even look into deprecating the `usingComparator` methods in favor of this new API for doing assertions on `Comparator` instances.  I can't imaging a case where you want to use a specific comparator for asserting comparison results without that comparator being the actual object under test.

>> Regarding the suggestion of registerComparable: as it is a static method, wouldn't we have the problem that it affects tests that come after the current test method?
>
> I understood this is what you wanted to have, but maybe I was wrong.

What I wanted was not having to call a specific method to pass the comparator on each assertion within a single test case.  But separate unit tests should by definition be independent of each other (even be able to run in parallel) so statically registering a comparator seems problematic, unless you know a way to have it impact only the current test and not those that will run afterwards or are running concurrently.
With the API I proposed above, this isn't really necessary anymore: the comparator is passed each time but without calling an additional method, and `isIncreasingOrder()` / `isDecreasingOrder()` already replace multiple separate assertions.

**Comment by scordio** [comment]
Thanks for the input @HermanBovens, you raised several interesting points which deserve to be discussed further.

However, as a first step, I would like to come up with something that doesn't require a breaking change and doesn't introduce a new pattern to perform assertions (e.g., having the comparator as assertion target).

I might have an idea how to solve your original example:

```java
assertThat(object1).usingComparator(mySpecialOrdering).isLessThan(object2);
```

with the restriction that this would work only for objects without a dedicated assertion class, i.e., covered by `ObjectAssert`.

I'll raise a PR about it to continue the discussion.

**Comment by obfischer** [comment]
Any update on this issue?

**Comment by scordio** [comment]
@obfischer I started working on it with #2086. As it's a bit of work and we are currently finalizing the release of 3.21.0, it will take a bit more before being ready but I see good chances for 3.22.0.

**Comment by parnmatt** [comment]
Is there any hope for this feature (or similar) to re-enter scope?
